collections/red_black_tree.ts

```diff
@@ -254,7 +254,17 @@ export class RedBlackTree<T> extends BinarySearchTree<T> {
    * Returns true if found and removed.
    */
   override remove(value: T): boolean {
-    const node = this.removeNode(value) as (RedBlackNode<T> | null);
+    const nodeToRemove = this.findNode(value) as (RedBlackNode<T> | null);
+
+    if (!nodeToRemove) {
+      return false;
+    }
+
+    const node = this.removeNode(nodeToRemove) as (
```

Comment:
From another comment, I mentioned changing removeNode to only return a successorNode if there is one. The following change to this function should get it so that it will still return true/false based on if the node was found, then it will only do the fixup if there is a successorNode that requires it.

```js
override remove(value: T): boolean {
  const node = this.findNode(value) as (RedBlackNode<T> | null);
  const successorNode = node && this.removeNode(node) as (RedBlackNode<T> | null);
  if (successorNode && !successorNode.red) {
    this.removeFixup(successorNode.parent, successorNode.left ?? successorNode.right);
  }
  return !!node;
}
```

Reply:
It seems like `RedBlackTree` really needs the removed node, and we can't just give it a `successorNode`. I'll preserve the logic as it was previously written. It also conceptually makes more sense to me that the `remove` function would return the node that was removed.

Reply:
+1 for remove returning the node that was removed and not the successorNode